src/db/read.rs

```diff
@@ -12,10 +20,62 @@ pub enum Whence {
     Hash(String),
 }
 
+pub enum SubscriptionInnerError {
+    InvalidUtf8(std::string::FromUtf8Error),
+}
+
+pub struct SubscriptionInner {
+    key_set: BTreeSet<Vec<u8>>,
+    scans: Vec<ScanOptions>,
+}
+
+impl SubscriptionInner {
+    pub fn new() -> Self {
+        SubscriptionInner {
+            key_set: BTreeSet::new(),
+            scans: Vec::new(),
+        }
+    }
+
+    pub fn record_get(&mut self, key: Vec<u8>) {
+        self.key_set.insert(key);
+    }
+
+    pub fn record_has(&mut self, key: Vec<u8>) {
+        self.key_set.insert(key);
+    }
+
+    pub fn record_scan(&mut self, scan_options: ScanOptions) {
+        self.scans.push(scan_options);
+    }
+
+    // TODO(arv): Just make the error String::FromUtf8Error
+    pub fn touched_keys(&self) -> Result<Vec<String>, SubscriptionInnerError> {
+        // TODO: Consider draining the vec. We might not need this data more
+        // than once depending on usage.
+        use SubscriptionInnerError::*;
+
+        self.key_set
+            .iter()
+            .map(|b| String::from_utf8(b.to_vec()))
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(InvalidUtf8)
+    }
+
+    pub fn touched_scans(&self) -> Vec<ScanOptions> {
+        // TODO: Consider draining the vec. We might not need this data more
+        // than once depending on usage.
+        self.scans.clone()
+    }
+}
+
+pub type Subscription = Arc<RwLock<SubscriptionInner>>;
```

Comment:
In theory we do not need locks because JS and Wasm is single threaded but I don't think Rust would allow that.

The following case needs "concurrent" access.

```ts
rep.subscribe(tx => 
  Promise.all([tx.get('a'), tx.get('b')])
);
```

I do not see how this could deadlock but it is always a good thing to keep in mind.

Reply:
At least some of the proliferation of locks in the rust code is due to rust not allowing inner mutability without them. Where in another language we wouldn't bother putting a lock around a thing that we know is not going to be accessed concurrently, rust makes us.